onnxruntime/core/providers/webnn/builders/model.cc

```diff
@@ -103,15 +127,34 @@ void Model::AllocateInputOutputBuffers() {
     const auto& input_info = input_output_info_.at(input);
     const auto input_shape = input_info.shape;
     const auto num_elements = SafeInt<size_t>(Product(input_shape));
-    wnn_inputs_.set(input,
-                    emscripten::val::global("Float32Array").new_(static_cast<const int>(num_elements)));
+    const auto data_type = input_info.data_type;
+    switch (data_type) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16:
+        wnn_inputs_.set(input,
+                        emscripten::val::global("Uint16Array").new_(static_cast<const uint16_t>(num_elements)));
+        break;
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+        wnn_inputs_.set(input,
+                        emscripten::val::global("Float32Array").new_(static_cast<const float>(num_elements)));
+        break;
+      default:
+        break;
+    }
   }
   for (const auto& output : outputs_) {
     const auto& output_info = input_output_info_.at(output);
     const auto output_shape = output_info.shape;
     const auto num_elements = SafeInt<size_t>(Product(output_shape));
-    wnn_outputs_.set(output,
-                     emscripten::val::global("Float32Array").new_(static_cast<const int>(num_elements)));
+    const auto data_type = output_info.data_type;
+    switch (data_type) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16:
+        wnn_outputs_.set(output,
+                         emscripten::val::global("Uint16Array").new_(static_cast<const uint16_t>(num_elements)));
+        break;
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+        wnn_outputs_.set(output,
+                         emscripten::val::global("Float32Array").new_(static_cast<const int>(num_elements)));
```

Comment:
Should this be `static_cast<const float>(` instead of `int`? Line 138 uses float.

Reply:
I think the `num_elements` should be used as is, because it is the result of casting to `size_t` and `emscripten::val::global("Float32Array").new_(num_elements)` should be fine.